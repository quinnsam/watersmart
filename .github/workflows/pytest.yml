# see: https://docs.github.com/en/actions/automating-builds-and-tests/building-and-testing-python

name: Pytest

on:
  push:
    branches: [ main ]
  pull_request:
    branches: [ main ]
  workflow_dispatch:
  schedule:
    - cron: "0 0 * * *"

jobs:
  build:

    runs-on: ubuntu-latest
    strategy:
      fail-fast: false
      matrix:
        python-version: ["3.12"]

    steps:
    - uses: actions/checkout@v4
    - name: Set up Python ${{ matrix.python-version }}
      uses: actions/setup-python@v5
      with:
        python-version: ${{ matrix.python-version }}
    - name: Install dependencies
      run: |
        python -m pip install --upgrade pip
        pip install -r requirements.test.txt
    - name: Test with pytest
      run: |
        pytest
    - name: Validate coverage
      run: |
        coverage json -q -o - | python3 .github/validate_coverage.py