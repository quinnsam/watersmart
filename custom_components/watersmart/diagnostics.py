"""Diagnostics support for WaterSmart."""

from typing import Any, cast

from homeassistant.components.diagnostics import async_redact_data
from homeassistant.config_entries import ConfigEntry
from homeassistant.const import CONF_PASSWORD, CONF_USERNAME
from homeassistant.core import HomeAssistant

from .coordinator import WaterSmartUpdateCoordinator

TO_REDACT = {
    CONF_USERNAME,
    CONF_PASSWORD,
}


async def async_get_config_entry_diagnostics(  # noqa: RUF029
    hass: HomeAssistant,  # noqa: ARG001
    entry: ConfigEntry,
) -> dict[str, Any]:
    """Return diagnostics for a config entry."""
    coordinator: WaterSmartUpdateCoordinator = entry.runtime_data.coordinator

    return cast(
        "dict[str, Any]",
        async_redact_data(
            {
                "entry": entry.as_dict(),
                "data": coordinator.data,
            },
            TO_REDACT,
        ),
    )
