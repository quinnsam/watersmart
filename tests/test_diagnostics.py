"""Test WaterSmart diagnostics."""

from homeassistant.core import HomeAssistant
import pytest
from pytest_homeassistant_custom_component.common import MockConfigEntry
from pytest_homeassistant_custom_component.components.diagnostics import (
    get_diagnostics_for_config_entry,
)
from pytest_homeassistant_custom_component.typing import ClientSessionGenerator
from syrupy.assertion import SnapshotAssertion
from syrupy.filters import props


@pytest.mark.usefixtures("init_integration")
async def test_entry_diagnostics(
    hass: HomeAssistant,
    mock_config_entry: MockConfigEntry,
    mock_watersmart_client,
    hass_client: ClientSessionGenerator,
    snapshot: SnapshotAssertion,
) -> None:
    """Test config entry diagnostics."""
    assert await get_diagnostics_for_config_entry(
        hass, hass_client, mock_config_entry
    ) == snapshot(exclude=props("entry_id", "created_at", "modified_at"))
