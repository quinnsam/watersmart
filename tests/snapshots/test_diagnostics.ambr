# serializer version: 1
# name: test_entry_diagnostics
  dict({
    'data': dict({
      'gallons_for_most_recent_full_day_key': dict({
        'attrs': dict({
          'related': list([
          ]),
        }),
        'state': 0,
      }),
      'gallons_for_most_recent_hour': dict({
        'attrs': dict({
          'related': list([
            dict({
              'gallons': 7.48,
              'start': '2024-06-19T19:00:00-07:00',
            }),
            dict({
              'gallons': 0,
              'start': '2024-06-19T20:00:00-07:00',
            }),
            dict({
              'gallons': 7.48,
              'start': '2024-06-19T21:00:00-07:00',
            }),
            dict({
              'gallons': 0,
              'start': '2024-06-19T22:00:00-07:00',
            }),
          ]),
          'start': '2024-06-19T22:00:00-07:00',
        }),
        'state': 0,
      }),
      'hourly': list([
        dict({
          'flags': None,
          'gallons': 7.48,
          'leak_gallons': 0,
          'read_datetime': 1718823600,
        }),
        dict({
          'flags': None,
          'gallons': 0,
          'leak_gallons': 0,
          'read_datetime': 1718827200,
        }),
        dict({
          'flags': None,
          'gallons': 7.48,
          'leak_gallons': 0,
          'read_datetime': 1718830800,
        }),
        dict({
          'flags': None,
          'gallons': 0,
          'leak_gallons': 0,
          'read_datetime': 1718834400,
        }),
      ]),
    }),
    'entry': dict({
      'data': dict({
        'host': 'test',
        'password': '**REDACTED**',
        'username': '**REDACTED**',
      }),
      'disabled_by': None,
      'discovery_keys': dict({
      }),
      'domain': 'watersmart',
      'minor_version': 1,
      'options': dict({
      }),
      'pref_disable_new_entities': False,
      'pref_disable_polling': False,
      'source': 'user',
      'subentries': list([
      ]),
      'title': 'Mock Title',
      'unique_id': None,
      'version': 1,
    }),
  })
# ---
